### PR TITLE
ui: vendor NPM modules

### LINKED ISSUE #14615: build: Yarn registry errors can fail the build
As described in #14610, build 207117 failed because `yarn install` saw a 503 error while trying to install dependencies. The relevant piece of the build log:

```
[18:12:27]	+ go-test-teamcity
[18:13:29]	+ build/builder.sh go generate ./pkg/...
[18:14:35]	error An unexpected error occurred: "https://registry.yarnpkg.com/@types/bytebuffer/-/bytebuffer-5.0.33.tgz: Request failed \"503 Service Unavailable\"".
[18:14:35]	make[1]: *** [cockroach/pkg/ui/yarn.installed] Error 1
```

I'm not aware of any other failures caused by the Yarn registry, but if this continues to be a problem, @dt suggested we look into some means of vendoring our Yarn dependencies. Yarn has an [offline cache](https://yarnpkg.com/blog/2016/11/24/offline-mirror/) where it looks in a local directory for dependency tarballs before hitting the network. Probably more hassle than it's worth at the moment, but something to keep in mind.

### LINKED ISSUE #21432: build: `make build` should work while offline
I tried running `make build` while offline today, and it failed due to the `build/node-run.sh -C ./pkg/ui yarn install` step of the build. I was able to work around it by ripping everything related to `YARN_INSTALLED_TARGET` out of the Makefile, and the resulting binary's UI worked just fine.

It's possible that it works normally and only needed the network because I changed between branches that had different versions of the UI code on them, but even so it'd be nice if there was an easier way to build cockroach while offline than having to do manual Makefile surgery.

### PR DESCRIPTION
**Pull request by benesch**
Vendor the NPM packages that our UI uses by checking them into a
submodule, [github.com/cockroachdb/yarn-vendored](https://github.com/cockroachdb/yarn-vendored). This prevents NPM or
Yarn registry flakiness from failing our CI builds, and also ensures
that a `git clone --recurse-submodules` is sufficient to build
CockroachDB offline.

The feature that enables this is Yarn's "offline mirror." The blog post
that introduced the feature also serves as its documentation:

    https://yarnpkg.com/blog/2016/11/24/offline-mirror/

To ensure the offline mirror is really used, we pass the '--offline'
flag to all Yarn invocations in the build system.

Fix #14615.
Really fix #21432.

Release note: None

### CODE REVIEW

**Comment by bdarnell** [comment]
LGTM, but there should be some documentation about how to add/update packages here. 

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27035#-:-LG22BRrDPfpbfs4e4-C:b-61ikdo)*

**Comment by couchand** [comment]
Thanks for keeping up with this stuff @benesch!  Please don't merge until @vilterp and I have had a chance to check it out, thanks.

**Comment by benesch** [comment]
Yeah, waiting on writing documentation until its proven that TeamCity can actually build with this.

**Comment by benesch** [comment]
Ok, TeamCity is happy! I've added instructions on how to manage dependencies in this new world.

File: pkg/ui/README.md
Comment (vilterp) [inline pkg/ui/README.md]: 😂

**Review by vilterp — APPROVED** [review]
Thanks for the fix and the thorough docs! Sad that the repo is so flaky that this is necessary.

@couchand pointed out that rebasing branches before and after this is going to be tricky, but might as well do it now.

**Comment by couchand** [comment]
I had to `make ui-maintainer-clean` to get this to build.  is that just me? if it's everyone that might deserve a shout out somewhere.

Otherwise LGTM

**Comment by benesch** [comment]
Hopefully just you! TeamCity and my GCE worker had no problem with it, for example. You didn't happen to save the error message, did you?

**Comment by couchand** [comment]
```
cp pkg/ui/node_modules/protobufjs/cli/{package.standalone.json,package.json}
ln -s ../../../yarn.protobufjs-cli.lock pkg/ui/node_modules/protobufjs/cli/yarn.lock
ln: failed to create symbolic link 'pkg/ui/node_modules/protobufjs/cli/yarn.lock': File exists
Makefile:334: recipe for target 'pkg/ui/yarn.installed' failed
```

**Comment by benesch** [comment]
Aha! Thanks. I know how to fix that.

**Comment by benesch** [comment]
Ok, fixed. Merging! TFTRs.

bors r=couchand,vilterp,bdarnell

**Pull request merged**
